examples/pandas/split-apply-combine/my_functions.py

```diff
@@ -0,0 +1,179 @@
+    """
+    Return a DataFrame with a series 'Tax Formula' from 'Income', 'Tax Rate' and 'Tax Credit' series.
+
+    :param income: the 'Income' series
+    :param tax_rate: the 'Tax Rate' series
+    :param tax_credit: the 'Tax Credit' series
+
+    :return: the DataFrame with the 'Tax Formula' Series
+    """
+    df = DataFrame({'income': income, 'tax_rate': tax_rate, 'tax_credit': tax_credit})
+    df["Tax Formula"] = df.apply(
+        lambda x: (
+            f"({int(x['income'])} * {x['tax_rate']})"
+            if np.isnan(x["tax_credit"]) else
+            f"({int(x['income'])} * {x['tax_rate']}) - ({int(x['income'])} * {x['tax_rate']}) * {x['tax_credit']}"
+        ), axis=1
+    )
+    return df["Tax Formula"]
+
+
+def tax(tax_formula: Series) -> Series:
+    """
+    Return a series 'Tax' from 'Tax Formula' series.
+    :param tax_formula: the 'Tax Formula' series.
+    :return: the 'Tax Formula' Series
+    """
+    df = tax_formula.to_frame()
+    df["Tax"] = df["Tax Formula"].apply(lambda x: round(pandas.eval(x)))
+    return df["Tax"]
+
+
+# ----------------------------------------------------------------------------------------------------------------------
+# Tax calculation pipelines (chained set of transformations)
+# ----------------------------------------------------------------------------------------------------------------------
+
+@pipe(
+    step(tax_rate, tax_rates=tax_rates_rules),  # apply the tax_rate step
+    step(tax_credit, tax_credits=tax_credits_rules),  # apply the tax_credit step
+)
+def under_100k_tax(under_100k: DataFrame) -> DataFrame:
+    """
+    Tax calculation pipeline for 'Income' under 100k.
+    :param df: The DataFrame  where 'Income' is under 100k
+    :return: the DataFrame with the 'Tax' Series
+    """
+    return under_100k
+
+
+@pipe(
+    step(tax_rate, tax_rates=tax_rates_rules),  # apply the tax_rate step
+)
+def over_100k_tax(over_100k: DataFrame) -> DataFrame:
+    """
+    Tax calculation pipeline for 'Income' over 100k.
+    :param over_100k: The DataFrame where 'Income' is over 100k
+    :return: the DataFrame with the 'Tax' Series
+    """
+    return over_100k
+
+
+# ----------------------------------------------------------------------------------------------------------------------
+# Tax calculator dataflow
+# ----------------------------------------------------------------------------------------------------------------------
+
+@extract_fields(
+    {"under_100k": DataFrame, "over_100k": DataFrame}
+)
+# This is the first node in the DAG
+def split_dataframe(input: DataFrame) -> Dict[str, DataFrame]:
+    """
+    That function takes the DataFrame in input and split it in 2 DataFrames:
+      - under_100k: Rows where 'Income' is under 100k
+      - over_100k: Rows where 'Income' is over 100k
+    """
+    return {
+        "under_100k": input.query('Income < 100000'),
+        "over_100k": input.query('Income >= 100000'),
+    }
```

Comment:
@nhuray for me I generally like to see files in rough topological order of execution. Here you've specifically grouped things. Any specific reasoning or thoughts behind how you've laid things out?

E.g. to me `split_dataframe` should be above `over_100k_tax` since that way I can read the function and then below see where it is used.